### DISCUSSION TITLE
Problems with body in a PUT operation

### DISCUSSION THREAD

**Discussion by Mat-moran** [vercel/next.js#29558] [Help] [answered]
Hi :smile: 

I am having problems with next reaching an end point to do a `PUT` request. With `GET` request I do not have problems as they have no body.

The problems seems to be some kind of JSON reformating in the body part because I can reach the API point and do the `PUT` operation without problems with pure React with a similar code (you can check it down below).

The backend server give me the next error when I do the request with next:
 `/shopinvader/customer/create: Invalid JSON data: Expecting value: line 1 column 1 (char 0)`
and the next log is 
`API resolved without sending a response for /api/customercreate, this may result in stalled requests.`

Any one knows what can be happening here? :pray:   Help 

The code I am using to reach the API is the next one:
### NEXT CODE IN pages/api/...
```
export default async (req, res) => {
  const customerEndPoint = process.env.SHOPINVADER_BASE_URL + "/customer/create";

  const fetchOptions = {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "API-KEY": process.env.SHOPINVADER_API_KEY,
      "WEBSITE-UNIQUE-KEY": process.env.SHOPINVADER_WEBSITE_UNIQUE_KEY,
    },
    body: JSON.stringify({
      "name": 'amadeio',
      "email": 'aaaa@gmail.com',
      'city': 'Anteuqera',
      'country': { 'id': '68' },
      'external_id': '186',
      'street': 'lucena',
      'zip': '29200',
      'is_company': "true"
    })
  }
  console.log(fetchOptions)

  const response = await fetch(
    customerEndPoint,
    fetchOptions,
  );
```
### REACT Code 
```
const customerCreate = async () => {
    const url = SHOPINVADER_BASE_URL + "/customer/create";

    const fetchOptions = {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "API-KEY": SHOPINVADER_API_KEY,
        "WEBSITE-UNIQUE-KEY": SHOPINVADER_WEBSITE_UNIQUE_KEY,
      },
      body: JSON.stringify({
        "name": 'amadeio',
        "email": 'aaaa@gmail.com',
        'city': 'Anteuqera',
        'country': { 'id': '68' },
        'external_id': '186',
        'street': 'lucena',
        'zip': '29200',
        'is_company': "true"
      })
    }
    console.log(fetchOptions)

    const response = await fetch(
      url,
      fetchOptions,
    );
  }

```

**Comment by zeekrey** [comment, marked as answer]
I don't get the whole problem. But, to resolve the stalled request issue, your api needs to send an actual response by using the response helpers: 

> res.status(code) - A function to set the status code. code must be a valid HTTP status code
> res.json(body) - Sends a JSON response. body must be a serialiazable object
> res.send(body) - Sends the HTTP response. body can be a string, an object or a Buffer
> res.redirect([status,] path) - Redirects to a specified path or URL. status must be a valid HTTP status code. If not specified, status defaults to "307" "Temporary redirect".
~ https://nextjs.org/docs/api-routes/response-helpers

In your case something like this:

```javascript
//...

res.send(response)
```

**Reply by Mat-moran** [reply]
Hi @zeekrey the problem is not with the response is throwing the __next api__ to the front end. The problem is with the backend that is other server and other api that I am trying to reach and do some staff. 

It is the backend that is throwing the error message `/shopinvader/customer/create: Invalid JSON data: Expecting value: line 1 column 1 (char 0)` when the POST method is launch and this does not happend when I do the same with pure React.

Also I correct the first post because I am using `JSON.stringify()` to pass the JSON object.

**Reply by zeekrey** [reply]
Hey @Mat-moran ,

I created a example repo: https://stackblitz.com/edit/nextjs-6awhz5?file=pages/index.js

I wrote a fetch function that is called from the client (within useEffect), from getStaticProps and from and API. Everything looks fine. There is probably an issue with your implementation or the json response.

**Reply by Mat-moran** [reply]
Thanks you very much @zeekrey you are a boss!

It should be a problem with my API. The extrange thing is that I tryed the same header and body with Reactjs and with Nextjs from an api, the Reactjs works fine and I am able to create and with the next it give me an error with the json it is receiving the API. crazy thing!! 
I will try to debug from my side what can be happening...

good day!

**Reply by Mat-moran** [reply]
Maybe I have something wrong in my previows implementation. I did the way proposed by @zeekrey and it works. I close it. Thanks men!!

**Reply by Mat-moran** [reply]
Also I notice that I was pointing to `http` instead of `htpps` that works for GET request but not for POST.
